riscv/mmu.h

```diff
@@ -206,15 +206,34 @@ class mmu_t
     return (float128_t){load<uint64_t>(addr), load<uint64_t>(addr + 8)};
   }
 
+  // AMO/Zicbom faults should be reported as store faults
+  // Zicbom instruction has different tval and starting address
+  #define convert_cbo_traps(BODY) \
```

Comment:
Having a second copy of these is quite unfortunate.  Can't we do this in a DRY way, e.g. by aligning `addr` to cache-line boundary as soon as you enter the `clean_inval` function?

Reply:
In other words, what's wrong with reporting `addr & -blocksz` in `tval`?  I realize it's a behavioral change, but is it actually wrong?

Reply:
We need the original address for the tval when raising exceptions.

Reply:
> We need the original address for the tval when raising exceptions.

Yeah, I'm questioning that. Is that really required?

Reply:
Sounds like this won't fly.  But I have a better idea.

The spec says "The PMP access control bits shall be the same for all physical addresses in the cache block" [... else] the behavior of a CBO instruction is UNSPECIFIED."

So, we don't need to check the entire CBO.  We only need to check the byte rs1 points to.  So, the following PR suffices:

```
diff --git a/riscv/mmu.h b/riscv/mmu.h
index 5a4835c3..efc6e9de 100644
--- a/riscv/mmu.h
+++ b/riscv/mmu.h
@@ -214,7 +214,7 @@ public:
 
   void clean_inval(reg_t addr, bool clean, bool inval) {
     convert_load_traps_to_store_traps({
-        const reg_t paddr = translate(generate_access_info(addr, LOAD, {false, false, false}), blocksz) & ~(blocksz - 1);
+      const reg_t paddr = translate(generate_access_info(addr, LOAD, {false, false, false}), 1);
       if (sim->reservable(paddr)) {
         if (tracer.interested_in_range(paddr, paddr + PGSIZE, LOAD))
           tracer.clean_invalidate(paddr, blocksz, clean, inval);
```

Reply:
Thank you for the suggestion. That cleans up the code elegantly.